### ISSUE TITLE
enum incorrectly narrowed to `Never` when instance check inside method

### ISSUE THREAD

**Issue by DetachHead** [microsoft/pyright#2299]
![image](https://user-images.githubusercontent.com/57028336/133196429-b9f3e02a-3454-468f-88da-20d3e819deae.png)
```python
from enum import Enum, auto


class Foo(Enum):
    bar = auto()
    baz = auto()

    def foo(self) -> None:
        if self is Foo.bar:
            reveal_type(self)  # Type of "self" is "Never"


def foo(value: Foo) -> None:
    if value is Foo.bar:
        reveal_type(value)  # Type of "value" is "Literal[Foo.bar]"
```

**VS Code extension or command-line**
vscode, pylance v2021.9.1

**Comment by DetachHead** [comment]
also why does pyright call it `Never`? shouldn't it be `NoReturn`?

**Comment by erictraut** [comment]
`Never` is effectively the same as `NoReturn`, but `NoReturn` has a specific meaning in Python that would be confusing in this context since most people associate it with the return type of a function that never returns (hence the name). There has been discussion in the typing-sig of officially exporting a `Never` symbol from `typing` because it better reflects what happens when a type is narrowed to an empty union. The name `Never` is borrowed from TypeScript, where it appears as a built-in type.

**Comment by erictraut** [comment]
I agree this is a bug in the `a is b` type narrowing logic. Thanks for reporting it.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.169. It will also be included in the next release of pylance.
